### ISSUE TITLE
SC2102 issues wrong warning in negated character class

### ISSUE THREAD

**Issue by rillig** [koalaman/shellcheck#1706]
#### For bugs
- SC2102
- online
- [x] The [rule's wiki page](https://shellcheck.net/wiki/SC2102) does not already cover this
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#! /bin/sh

case $1 in
[!!]) echo wrong
esac
```

#### Here's what shellcheck currently says:

~~~none
Line 4:
[!!]) echo wrong
^-- SC2102: Ranges can only match single chars (mentioned due to duplicates).
~~~

#### Here's what I wanted or expected to see:

No warning since the first `!` is for negation, and the second `!` is literal.

**Comment by koalaman** [comment]
Nice catch! This has now been fixed, thanks!

**Comment by rillig** [comment]
According to [POSIX](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_13_01), only the `!` character means negation:

> A bracket expression starting with an unquoted <circumflex> character produces unspecified results.

This could be reason enough to provide another warning. I haven't researched which shells support it.

**Comment by willholen** [comment]
Such a warning is already in place when you use `#!/bin/sh`

**Comment by bsdshell** [comment]
Hi,
how to disable all the warnings? 
can I just enable all the errors? 
shellcheck on SpaceEmacs.

**Comment by koalaman** [comment]
@bsdshell You can use `shellcheck -S error`. I don't know how you'd set this up in SpaceEmacs, but shellcheck will at least look for an environment variable `export SHELLCHECK_OPTS='-S error'`
